File: 프로그래머스/2/389479. 서버 증설 횟수/서버 증설 횟수.py
from heapq import heappush, heappop


def solution(players, m, k):
    answer = 0
    q = []
    
    heappush(q, 24)
    
    for t, p in enumerate(players):
        while q and q[0] == t:
            heappop(q)
        
        if p < len(q) * m:
            continue
    
        while len(q) * m <= p:
            heappush(q, t + k)
            answer += 1
    
    return answer

File: 프로그래머스/2/389479. 서버 증설 횟수/서버 증설 횟수.java
import java.util.PriorityQueue;

class Solution {

    public int solution(int[] players, int m, int k) {
        int answer = 0;

        PriorityQueue<Integer> pq = new PriorityQueue<>();
        pq.offer(24);

        for (int t = 0; t < players.length; t++) {
            while (!pq.isEmpty() && pq.peek() == t) {
                pq.poll();
            }

            while (players[t] >= pq.size() * m) {
                pq.offer(t + k);
                answer++;
            }
        }

        return answer;
    }
}